haystack/preview/components/samplers/top_p.py

```diff
@@ -0,0 +1,142 @@
+import logging
+from typing import List, Optional, Dict, Any
+
+from haystack.preview import ComponentError
+
+from haystack.lazy_imports import LazyImport
+from haystack.preview import Document, component, default_from_dict, default_to_dict
+
+
+logger = logging.getLogger(__name__)
+
+
+with LazyImport(message="Run 'pip install transformers[torch,sentencepiece]==4.32.1'") as torch_and_transformers_import:
+    import torch
+
+
+@component
+class TopPSampler:
+    """
+    Selects documents based on the cumulative probability of the Document similarity scores using top p (nucleus)
+    sampling.
+
+    Usage example:
+    ```
+    from haystack.preview import Document
+    from haystack.preview.components.samplers import TopPSampler
+
+    sampler = TopPSampler(top_p=0.95)
+    docs = [
+        Document(text="Berlin", metadata={"similarity_score": -10.6}),
+        Document(text="Belgrade", metadata={"similarity_score": -8.9}),
+        Document(text="Sarajevo", metadata={"similarity_score": -4.6}),
+    ]
+    query = "City in Bosnia and Herzegovina"
+    output = sampler.run(documents=docs)
+    docs = output["documents"]
+    assert len(docs) == 1
+    assert docs[0].text == "Sarajevo"
+    ```
+    """
+
+    def __init__(self, top_p: Optional[float] = 1.0, score_field: Optional[str] = "similarity_score"):
+        """
+        Creates an instance of TopPSampler.
+
+        :param top_p: Cumulative probability threshold for filtering the documents (usually between 0.9 and 0.99).
+        :param score_field: The name of the field that should be used to store the scores in a document's metadata.
+        """
+        torch_and_transformers_import.check()
+        super().__init__()
+
+        self.top_p = top_p
+        self.score_field = score_field
+        self.model = None
+        self.tokenizer = None
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        """
+        return default_to_dict(self, top_p=self.top_p, score_field=self.score_field)
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "TopPSampler":
+        """
+        Deserialize this component from a dictionary.
+        """
+        return default_from_dict(cls, data)
+
+    @component.output_types(documents=List[Document])
+    def run(self, documents: List[Document], top_p: Optional[float] = None):
```

Comment:
What's the use case of having run take `top_p`? Would it make more sense to have it only in `__init__` or only in `run`?

Reply:
I know where striving for simplicity in this phase but this is a major important param for runtime